doc/scs2ascii.1
'\" t
.ig
Man page for scs2ascii.

Copyright (C) 1999 - 2008 Carey Evans.

You can redistribute and/or modify this document under the terms of 
the GNU General Public License as published by the Free Software
Foundation; either version 2 of the License, or (at your option)
any later version.

This document is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
..
.TH SCS2ASCII 1 "17 Jun 2001"
.SH NAME
scs2ascii \- convert IBM SCS printer data into ASCII
.SH SYNOPSIS
.B scs2ascii
.SH DESCRIPTION
This program handles the conversion from an SCS print stream from an
AS/400 printer emulation to something suitable for sending to lpr.
It is only intended for use with
.BR lp5250d .
.PP
When using host print transform with
.BR lp5250d ,
.B scs2ascii
will remove the remaining characters from the SCS print stream to
leave only the data intended for the printer.
.SH "SEE ALSO"
.BR lp5250d (1),
.B https://tn5250.github.io/
.SH COPYRIGHT
.B scs2ascii
is copyright
.if t \(co
1997 \- 1999 Michael Madore.  This manpage is copyright
.if t \(co
1999 \- 2008 Carey Evans.
.PP
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation; either version 2.1 of the License, or
(at your option) any later version.
.PP
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.
.PP
You should have received a copy of the GNU Lesser General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
.SH AUTHORS
.B scs2ascii
was written by Michael Madore.
This manual page was written by Carey Evans.


NEWS
tn5250 NEWS
------------------------------------------------------------------------------

Wed May 04, 2005:
	* The tn5250 project is hosted on sourceforge at:
		http://tn5250.sourceforge.net
	* There is a mailing list and archives at David Gibbs midrange site:
		http://www.midrange.com
	  The mailing list is at:
		http://lists.midrange.com/mailman/listinfo/linux5250
	* There is an X11 front end written by James Rich.  The site is:
		http://www.chowhouse.com/~james/x5250
	* For all other news please visit the project page on sourceforge.

Mon Dec 06, 1999:
	* A FAQ is now in existence, thanks to Philippe Petrinko.  You can
	  browse the FAQ at http://perso.libertysurf.fr/plinux/tn5250-faq.html
	* The CVS server is now serving up the mailing list archives, thanks
	  to David Gibbs.   You can browse the archives at
	  http://cvs.cleveland.lug.net/mail-archives/tn5250



lp5250d/scs2ascii.c
/* TN5250 - An implementation of the 5250 telnet protocol.
 * Copyright (C) 1997-2008 Michael Madore
 * 
 * This file is part of TN5250.
 *
 * TN5250 is free software; you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation; either version 2.1, or (at your option)
 * any later version.
 * 
 * TN5250 is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public License
 * along with this software; see the file COPYING.  If not, write to
 * the Free Software Foundation, Inc., 59 Temple Place, Suite 330,
 * Boston, MA 02111-1307 USA
 * 
 */

#include "tn5250-private.h"
#include "scs-private.h"

/*
#define DEBUG
*/


static void scs2ascii_pp (Tn5250SCS * This);
static void scs2ascii_ahpp (int *curpos);
void scs2ascii_avpp (Tn5250SCS * This);
void scs2ascii_transparent (Tn5250SCS * This);
void scs2ascii_ff (Tn5250SCS * This);
void scs2ascii_nl (Tn5250SCS * This);
void scs2ascii_default (Tn5250SCS * This);
Tn5250SCS *tn5250_scs2ascii_new ();

Tn5250CharMap *map;

int
main ()
{
  Tn5250SCS *scs = NULL;

  if ((getenv ("TN5250_CCSIDMAP")) != NULL)
    {
      map = tn5250_char_map_new (getenv ("TN5250_CCSIDMAP"));
    }
  else
    {
      map = tn5250_char_map_new ("37");
    }


  /* Initialize the scs toolkit */
  scs = tn5250_scs2ascii_new ();

  if (scs == NULL)
    {
      return (-1);
    }

  /* And now set up our callbacks */
  scs->column = 1;
  scs->row = 1;


  /* Turn control over to the SCS toolkit and run the event loop */
  scs_main (scs);

  tn5250_char_map_destroy (map);
  free (scs);
  return (0);
}




/* This initializes the scs callbacks
 */
Tn5250SCS *
tn5250_scs2ascii_new ()
{
  Tn5250SCS *scs = tn5250_scs_new ();

  if (scs == NULL)
    {
      fprintf (stderr,
	       "Unable to allocate memory in tn5250_scs2ascii_new ()!\n");
      return NULL;
    }

  /*
     scs->data = tn5250_new(struct _Tn5250SCSPrivate, 1);
     if (scs->data == NULL) {
     free(scs);
     return;
     }
   */

  /* And now set up our callbacks */
  scs->transparent = scs2ascii_transparent;
  scs->ff = scs2ascii_ff;
  scs->rff = scs2ascii_ff;
  scs->nl = scs2ascii_nl;
  scs->rnl = scs2ascii_nl;
  scs->pp = scs2ascii_pp;
  scs->avpp = scs2ascii_avpp;
  scs->scsdefault = scs2ascii_default;
  return scs;
}



void
scs2ascii_default (Tn5250SCS * This)
{
#ifdef DEBUG
#ifdef VERBOSE
  fprintf (stderr, "doing scs2ascii_default()\n");
#endif
#endif
  printf ("%c", tn5250_char_map_to_local (map, This->curchar));
  This->column++;
#ifdef DEBUG
#ifdef VERBOSE
  fprintf (stderr, "%c (%x)\n",
	   tn5250_char_map_to_local (map, This->curchar), This->curchar);
#endif
#endif
  return;
}

void
scs2ascii_ff (Tn5250SCS * This)
{
#ifdef DEBUG
#ifdef VERBOSE
  fprintf (stderr, "doing scs2ascii_ff()\n");
#endif
#endif
  scs_ff (This);
  printf ("\f");
  return;
}

void
scs2ascii_nl (Tn5250SCS * This)
{
#ifdef DEBUG
#ifdef VERBOSE
  fprintf (stderr, "doing scs2ascii_nl()\n");
#endif
#endif
  scs_nl (This);
  printf ("\n");
  return;
}

static void
scs2ascii_pp (Tn5250SCS * This)
{
  unsigned char curchar;

#ifdef DEBUG
#ifdef VERBOSE
  fprintf (stderr, "doing scs2ascii_pp()\n");
#endif
#endif
  curchar = fgetc (stdin);
  switch (curchar)
    {
    case SCS_AVPP:
      {
	This->avpp (This);
	break;
      }
    case SCS_AHPP:
      {
	scs2ascii_ahpp (&(This->column));
	break;
      }
    default:
      {
	fprintf (stderr, "ERROR: Unknown 0x34 command %x\n", curchar);
      }
    }
}

void
scs2ascii_avpp (Tn5250SCS * This)
{
  int i;
  int newrow;
  int lines;

#ifdef DEBUG
#ifdef VERBOSE
  fprintf (stderr, "doing scs2ascii_avpp()\n");
#endif
#endif
  newrow = fgetc (stdin);
#ifdef DEBUG
  fprintf (stderr, "AVPP %d\n", newrow);
#endif

  if (newrow < This->row)
    {
      printf ("\f");
      This->row = 1;
    }
  else
    {
      lines = newrow - This->row;

      for (i = 0; i < lines; i++)
	{
	  printf ("\n");
	}

      This->row = newrow;
    }
  return;
}

static void
scs2ascii_ahpp (int *curpos)
{
  int position;
  int loop;

#ifdef DEBUG
#ifdef VERBOSE
  fprintf (stderr, "doing scs2ascii_ahpp()\n");
#endif
#endif
  position = fgetc (stdin);
#ifdef DEBUG
  fprintf (stderr, "AHPP %d (current position: %d)\n", position, *curpos);
#endif

  if (*curpos > position)
    {
      printf ("\r");
      for (loop = 0; loop < position; loop++)
	{
	  printf (" ");
	}
    }
  else
    {
      for (loop = 0; loop < position - *curpos; loop++)
	{
	  printf (" ");
	}
    }
  *curpos = position;
  return;
}

void
scs2ascii_transparent (Tn5250SCS * This)
{

  int bytecount;
  int loop;

#ifdef DEBUG
#ifdef VERBOSE
  fprintf (stderr, "doing scs2ascii_transparent()\n");
#endif
#endif
  bytecount = fgetc (stdin);
#ifdef DEBUG
  fprintf (stderr, "TRANSPARENT (%x) = ", bytecount);
#endif

  for (loop = 0; loop < bytecount; loop++)
    {
      printf ("%c", fgetc (stdin));
    }
}

/* vi:set sts=3 sw=3: */
